src/Config/DataSource.cs

```diff
@@ -1,127 +1,50 @@
+        {
+            return (TOptionType)(object)new MsSqlOptions(
+                SetSessionContext: ReadBoolOption(namingPolicy.ConvertName(nameof(MsSqlOptions.SetSessionContext))));
         }
-    }
 
-    /// <summary>
-    /// Options for cosmosdb_nosql database.
-    /// </summary>
-    public record CosmosDbNoSqlOptions(
-        string Database,
-        string? Container,
-        [property: JsonPropertyName(CosmosDbNoSqlOptions.GRAPHQL_SCHEMA_PATH_PROPERTY_NAME)]
-        string? GraphQLSchemaPath,
-        [property: JsonIgnore]
-        string? GraphQLSchema)
-    {
-        public const string GRAPHQL_SCHEMA_PATH_PROPERTY_NAME = "schema";
+        throw new NotImplementedException();
     }
 
-    /// <summary>
-    /// Options for MsSql database.
-    /// </summary>
-    public record MsSqlOptions(
-        [property: JsonPropertyName("set-session-context")]
-        bool SetSessionContext = false)
-    {
-        public const string JSON_PROPERTY_NAME = nameof(DatabaseType.mssql);
-
-        public MsSqlOptions()
-            : this(SetSessionContext: true) { }
-    }
+    private string? ReadStringOption(string option) => Options.ContainsKey(option) ? Options[option].GetString() : null;
+    private bool ReadBoolOption(string option) => Options.ContainsKey(option) ? Options[option].GetBoolean() : false;
 
-    /// <summary>
-    /// Options for PostgresSql database.
-    /// </summary>
-    public record PostgreSqlOptions
-    {
-        public const string JSON_PROPERTY_NAME = nameof(DatabaseType.postgresql);
-    }
+    [JsonIgnore]
+    public string DatabaseTypeNotSupportedMessage => $"The provided database-type value: {DatabaseType} is currently not supported. Please check the configuration file.";
+}
 
-    /// <summary>
-    /// Options for CosmosDb_PostgresSql database.
-    /// </summary>
-    public record CosmosDbPostgreSqlOptions { }
+public interface IDataSourceOptions { }
 
-    /// <summary>
-    /// Options for MySql database.
-    /// </summary>
-    public record MySqlOptions
-    {
-        public const string JSON_PROPERTY_NAME = nameof(DatabaseType.mysql);
-    }
+/// <summary>
+/// The CosmosDB NoSQL connection options.
+/// </summary>
+/// <param name="Database">Name of the default CosmosDB database.</param>
+/// <param name="Container">Name of the default CosmosDB container.</param>
+/// <param name="Schema">Path to the GraphQL schema file.</param>
+/// <param name="GraphQLSchema">Raw contents of the GraphQL schema.</param>
+public record CosmosDbNoSQLDataSourceOptions(string? Database, string? Container, string? Schema, string? GraphQLSchema) : IDataSourceOptions;
 
-    /// <summary>
-    /// Enum for the supported database types.
-    /// </summary>
-    public enum DatabaseType
-    {
-        cosmosdb_postgresql,
-        cosmosdb_nosql,
-        mssql,
-        mysql,
-        postgresql
-    }
-}
+public record MsSqlOptions(bool SetSessionContext = true) : IDataSourceOptions;
```

Comment:
For context, here's the previous implementation:

```c#
    public record MsSqlOptions(
        [property: JsonPropertyName("set-session-context")]
        bool SetSessionContext = false)
    {
        public const string JSON_PROPERTY_NAME = nameof(DatabaseType.mssql);

        public MsSqlOptions()
            : this(SetSessionContext: true) { }
    }
```

On the `record` definition we initialise the `SetSessionContext` property to `false`, but then there is a default constructor that provides `true` to the property.

Everything that I was able to figure out from how deserialization works, object construction works, breaking down the IL and asking people is that the default constructor will **always** get called, meaning that the property will always get initialised with `true`, since you simply can't create it without invoking the default constructor.

That's why I simplified it as I did.